File: packages/typescript/test/intelx.test.ts
import { describe, it, expect, vi } from "vitest";
import { LeaksNowClient } from "../src/client.js";

function jsonResponse(body: unknown) {
  return new Response(JSON.stringify(body), { status: 200, headers: { "content-type": "application/json" } });
}

describe("client.intelx", () => {
  it("downloads lists history", async () => {
    const fetchMock = vi.fn().mockResolvedValue(jsonResponse({ items: [] }));
    const client = new LeaksNowClient("ms_key", { fetch: fetchMock });
    await client.intelx.downloads();
    expect(fetchMock.mock.calls[0][0]).toBe("https://leaks.now/api/v1/intelx/downloads");
    expect(fetchMock.mock.calls[0][1].method).toBe("GET");
  });

  it("download POSTs systemId+bucket", async () => {
    const fetchMock = vi.fn().mockResolvedValue(jsonResponse({ id: "1" }));
    const client = new LeaksNowClient("ms_key", { fetch: fetchMock });
    await client.intelx.download({ systemId: "52c30047-3a62-41c8-bb1b-769520b90957", bucket: "leaks.private.general" });
    expect(fetchMock.mock.calls[0][0]).toBe("https://leaks.now/api/v1/intelx/download");
    expect(JSON.parse(fetchMock.mock.calls[0][1].body)).toEqual({
      systemId: "52c30047-3a62-41c8-bb1b-769520b90957",
      bucket: "leaks.private.general",
    });
  });

  it("getFile returns binary with content-type and filename", async () => {
    const bytes = new Uint8Array([1, 2, 3, 4]);
    const res = new Response(bytes, {
      status: 200,
      headers: {
        "content-type": "application/octet-stream",
        "content-disposition": 'attachment; filename="dump.bin"',
      },
    });
    const fetchMock = vi.fn().mockResolvedValue(res);
    const client = new LeaksNowClient("ms_key", { fetch: fetchMock });
    const file = await client.intelx.getFile(7);

    expect(fetchMock.mock.calls[0][0]).toBe("https://leaks.now/api/v1/intelx/downloads/7/file");
    expect(file.contentType).toBe("application/octet-stream");
    expect(file.filename).toBe("dump.bin");
    expect(new Uint8Array(file.data)).toEqual(bytes);
  });

  it("deleteDownload DELETEs by id", async () => {
    const fetchMock = vi.fn().mockResolvedValue(jsonResponse({ ok: true }));
    const client = new LeaksNowClient("ms_key", { fetch: fetchMock });
    await client.intelx.deleteDownload(7);
    expect(fetchMock.mock.calls[0][0]).toBe("https://leaks.now/api/v1/intelx/downloads/7");
    expect(fetchMock.mock.calls[0][1].method).toBe("DELETE");
  });
});

File: go/intelx_test.go
package leaksnow

import (
	"bytes"
	"context"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestIntelX(t *testing.T) {
	var path, body, method string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		path = r.URL.Path
		method = r.Method
		b, _ := io.ReadAll(r.Body)
		body = string(b)
		_, _ = w.Write([]byte(`{"ok":true}`))
	}))
	defer srv.Close()
	c := NewClient("ms_key", WithBaseURL(srv.URL))
	ctx := context.Background()

	if _, err := c.IntelX.Downloads(ctx); err != nil {
		t.Fatal(err)
	}
	if path != "/api/v1/intelx/downloads" || method != http.MethodGet {
		t.Fatalf("downloads: %s %s", method, path)
	}

	if _, err := c.IntelX.Download(ctx, IntelXDownloadRequest{SystemID: "uuid-1", Bucket: BucketLeaksPrivateGeneral}); err != nil {
		t.Fatal(err)
	}
	if path != "/api/v1/intelx/download" {
		t.Fatalf("download path = %q", path)
	}
	if body != `{"systemId":"uuid-1","bucket":"leaks.private.general"}` {
		t.Fatalf("download body = %q", body)
	}

	if _, err := c.IntelX.DeleteDownload(ctx, "7"); err != nil {
		t.Fatal(err)
	}
	if path != "/api/v1/intelx/downloads/7" || method != http.MethodDelete {
		t.Fatalf("deleteDownload: %s %s", method, path)
	}
}

func TestIntelXGetFile(t *testing.T) {
	want := []byte{1, 2, 3, 4}
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/intelx/downloads/7/file" {
			t.Errorf("path = %q", r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/octet-stream")
		w.Header().Set("Content-Disposition", `attachment; filename="dump.bin"`)
		_, _ = w.Write(want)
	}))
	defer srv.Close()
	c := NewClient("ms_key", WithBaseURL(srv.URL))
	f, err := c.IntelX.GetFile(context.Background(), "7")
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(f.Data, want) {
		t.Fatalf("data = %v, want %v", f.Data, want)
	}
	if f.ContentType != "application/octet-stream" {
		t.Fatalf("contentType = %q", f.ContentType)
	}
	if f.Filename != "dump.bin" {
		t.Fatalf("filename = %q", f.Filename)
	}
}

func TestParseFilename(t *testing.T) {
	cases := map[string]string{
		`attachment; filename="dump.bin"`:                         "dump.bin",
		`attachment; filename=plain.bin`:                          "plain.bin",
		`attachment; filename*=UTF-8''r%C3%A9sum%C3%A9.bin`:       "résumé.bin",
		`attachment; filename="x.bin"; filename*=UTF-8''y%20.bin`: "y .bin",
		`attachment; filename*=UTF-8'en'file.bin`:                 "file.bin",
		``: "",
	}
	for in, want := range cases {
		if got := parseFilename(in); got != want {
			t.Errorf("parseFilename(%q) = %q, want %q", in, got, want)
		}
	}
}
